### PR TITLE
drivers: i2c+i3c: Place API into iterable section

### PR DESCRIPTION
**Pull request by pdgendt**
Add wrapper DEVICE_API macro to all i2c_driver_api & i3c_driver_api instances.

### CODE REVIEW

File: include/zephyr/drivers/i2c.h [hunk 1]
```diff
@@ -746,13 +743,11 @@ __syscall int i2c_get_config(const struct device *dev, uint32_t *dev_config);
...
 {
-	const struct i2c_driver_api *api = (const struct i2c_driver_api *)dev->api;
-
-	if (api->get_config == NULL) {
+	if (DEVICE_API_GET(i2c, dev)->get_config == NULL) {
 		return -ENOSYS;
 	}
 
-	return api->get_config(dev, dev_config);
+	return DEVICE_API_GET(i2c, dev)->get_config(dev, dev_config);
```
Comment (fabiobaltieri) [inline include/zephyr/drivers/i2c.h:750]: wouldn't you be better off keeping the `api` local variable in this case? Up to the maintainers really but I'd rather do

```c
	const struct i2c_driver_api *api = DEVICE_API_GET(i2c, dev);
```

if there's extra conditoins like null checks and not just the api call.

Anyway, up to you folks.

Comment (pdgendt) [reply include/zephyr/drivers/i2c.h]: Reverted the header changes for now as discussed in discord.

**Review by henrikbrixandersen — DISMISSED** [review]

**Comment by pdgendt** [comment]
Removed some leftover asserts.

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by pdgendt** [comment]
@teburd PTAL

**Review by fabiobaltieri — APPROVED** [review]

**Review by ydamigos — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]

**Review by XenuIsWatching — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Pull request merged**